python/ray/serve/_private/router.py

```diff
@@ -437,60 +436,60 @@ def update_deployment_config(self, deployment_config: DeploymentConfig):
     async def _resolve_deployment_responses(
         self, request_args: Tuple[Any], request_kwargs: Dict[str, Any]
     ) -> Tuple[Tuple[Any], Dict[str, Any]]:
-        """Replaces `DeploymentResponse` objects with their resolved object refs.
+        """Replaces top-level `DeploymentResponse` objects with resolved object refs.
 
-        Uses the `_PyObjScanner` to find and replace the objects. This
-        enables composition without explicitly calling `_to_object_ref`.
+        This enables composition without explicitly calling `_to_object_ref`.
         """
-        from ray.serve.handle import (
-            DeploymentResponse,
-            DeploymentResponseGenerator,
-            _DeploymentResponseBase,
-        )
+        from ray.serve.handle import DeploymentResponse, DeploymentResponseGenerator
+
+        new_args = [None for _ in range(len(request_args))]
+        new_kwargs = {}
+
+        arg_tasks = []
+        response_indices = []
+        for i, obj in enumerate(request_args):
+            if isinstance(obj, DeploymentResponseGenerator):
+                raise RuntimeError(
+                    "Streaming deployment handle results cannot be passed to "
+                    "downstream handle calls. If you have a use case requiring "
+                    "this feature, please file a feature request on GitHub."
+                )
+            elif isinstance(obj, DeploymentResponse):
+                # Launch async task to convert DeploymentResponse to an object ref, and
+                # keep track of the argument index in the original `request_args`
+                response_indices.append(i)
+                arg_tasks.append(asyncio.create_task(obj._to_object_ref()))
+            else:
+                new_args[i] = obj
+
+        kwarg_tasks = []
+        response_keys = []
+        for k, obj in request_kwargs.items():
+            if isinstance(obj, DeploymentResponseGenerator):
+                raise RuntimeError(
+                    "Streaming deployment handle results cannot be passed to "
+                    "downstream handle calls. If you have a use case requiring "
+                    "this feature, please file a feature request on GitHub."
+                )
```

Comment:
nit: share this error message to avoid copypasta

Reply:
done